README.md
# Sirius
Sirius 是用于Jenkins持续集成的常用功能集合，便于简单、快速的使用Jenkins实践持续集成。

## 文档

详细文档请参考[这里](http://scmesos06/docs/dfis/sirius/)

## Dependencies
- Linux System
- Python 2.7+
- Docker 1.8.3 (某些功能依赖)

## Quick Start
### Installation
sirius可以使用pip来快速安装，执行以下命令来快速安装sirius:

```shell
sudo pip --trusted-host scmesos06 install -i http://scmesos06/simple sirius

```

### Usage
我们在底层使用了[Fabric](http://docs.fabfile.org/en/1.11/index.html#)，所以我们借鉴Fabric的Task的概念。

#### Available Task
所以你可以执行下面有哪些可以用的Task：

```shell
sirius -l
# or
sirius --list 

```

执行完了之后， 你可以得到如下结果：

```shell
Available commands:

    docker_deploy      deploy a docker image on some server
    docker_image_name  get building docker image name
    usage              help information
    version            version information
        
```

#### Task Description

如果你想查看Task的详细信息，可以执行以下命令查看：
```shell
sirius -d docker_deploy
# or
sirius --display docker_deploy

```

那么我们就会得到docker_deploy 任务的详细信息:

```shell
Displaying detailed information for task 'docker_deploy':

    deploy a docker image on some server
    
        will create container when if container is not exists, otherwise update container
        Example:
            fab deploy:meerkat,meerkat:0.0.1,server=scdfis01,ports="3141;8080;8900;8081",env="DEBUG\=1;PATH\=2"
    
        :param name: container name
        :param image: image with tag, like: 'CentOS:7.0'
        :param server: which server will deploy the container, default:localhost
        :param ports: just one port, will mapping <port>:80, also accept list with public, private pairs, like: public1;private1;public2;private2
        :param volumes: like: host_file1;container_file1;host_file2;container_file2
        :param env: var=10;DEBUG=true
        :param cmd: `class`:`str`
        :param hostname:
        :return:
    
    Arguments: name, image, server=None, ports=None, volumes=None, env=None, cmd='', hostname='sirius'

```

#### Execute Task
我们得到了Task的详细信息之后，我们就可以根据描述信息来尝试使用， sirius采用这样 ``sirius COMMAND:param1,param2,param3=value3``的格式执行。
例如我们我们要执行部署工作，我们就可以执行如下命令：

```shell
sirius docker_deploy:container_name,docker.neg/meerkat:0.0.1,scdfis00, ports=3141;8080

```
通过这个命令我们就在scdfis00的服务器上使用docker.neg/meerka:0.0.1的镜像启动了一个叫做container_name的容器，
并把scdfis00 上的3141 tcp端口映射到了容器内部的8080tcp端口。当然还可以设置容器的卷、环境变量等，具体设置方式可以参照任务描述信息。

但是有一点必须注意的是， scdfis00上必须安装[dockerapi](http://trgit2/backend_framework/docker-manage-api)。


CHANGELOG.md
# Change log

## [0.0.4]
### Fixed
- fix jsonschema requirements missing
### Added
- add environment key for matrix version
- add default matrix version

## [0.0.3]

- add docker dfis prd deploy

## [0.0.3]

### Added
- add docker dev deploy

## [0.0.2]

### Added
- release tag for docker image
- support for windows

## [0.0.1]

### Added
- deploy docker command
- retrieve image name
- console entry point
- add build docker command



sirius/docker.py
from __future__ import print_function

import httplib
import json
import os
from distutils.version import LooseVersion
from itertools import chain
from itertools import imap
from etcd import Client

import requests
from fabric.api import local
from git import Repo
from simplekit import ContainerNotFound
from simplekit.docker import factory

from .setttings import DEFAULT_MATRIX_VERSION
from .setttings import DEIMOS
from .utils import group_by_2
from .utils import parse_list


def docker_dfis_prd_deploy(name, image, replicas=2, volumes=None, env=None, cmd="", hostname="sirius", servers=None):
    """deploy a docker image on dfis prd server

        will create container when if container is not exists, otherwise update container
        Example:
            sirius docker_dfis_prd_deploy:meerkat,meerkat:0.0.1,env="DEBUG\=1;PATH\=2",servers="s7dfis01;s7dfis02"

        :param name: container name
        :param image: image with tag, like: 'CentOS:7.0'
        :param replicas: container replicas
        :param volumes: like: host_file1;container_file1;host_file2;container_file2
        :param env: var=10;DEBUG=true
        :param cmd: `class`:`str`
        :param hostname:
        :param servers: s7dfis01;s7dfis02
        :return:
    """

    if servers is None:
        servers = ["s7dfis10", "s7dfis11", "s7dfis12", "s7dfis13"]
    elif servers:
        servers = parse_list(servers)

    project_name = name
    replicas = int(replicas)
    if replicas <= 0:
        raise Exception("replicas must more than 0")

    if env:
        env = parse_list(env)

    __deploy(project_name, image, replicas, volumes, env, cmd, hostname, servers, etcd_port=4007)


def docker_dev_deploy(name, image, replicas=1, volumes=None, env=None, cmd="", hostname="sirius", server=None):
    """deploy a docker image on dev server

        will create container when if container is not exists, otherwise update container
        Example:
            sirius docker_dev_deploy:meerkat,meerkat:0.0.1,env="DEBUG\=1;PATH\=2"

        :param name: container name
        :param image: image with tag, like: 'CentOS:7.0'
        :param replicas: container replicas
        :param volumes: like: host_file1;container_file1;host_file2;container_file2
        :param env: var=10;DEBUG=true
        :param cmd: `class`:`str`
        :param hostname:
        :param server:
        :return:
    """
    project_name = name
    replicas = int(replicas)
    if replicas <= 0:
        raise Exception("replicas must more than 0")

    if server is None:
        server = '10.16.78.82'

    if env:
        env = parse_list(env)
        if "ENV=gqc" in env:
            server = "10.1.24.134"

    __deploy(project_name, image, replicas, volumes, env, cmd, hostname, [server])


def docker_deploy(name, image, server=None, ports=None, volumes=None, env=None, cmd="", hostname="sirius"):
    """deploy a docker image on some server

    will create container when if container is not exists, otherwise update container
    Example:
        sirius docker_deploy:meerkat,meerkat:0.0.1,server=scdfis01,ports="3141;8080;8900;8081",env="DEBUG\=1;PATH\=2"

    :param name: container name
    :param image: image with tag, like: 'CentOS:7.0'
    :param server: which server will deploy the container, default:localhost
    :param ports: just one port, will mapping <port>:80, also accept list with public, private pairs,
           like: public1;private1;public2;private2
    :param volumes: like: host_file1;container_file1;host_file2;container_file2
    :param env: var=10;DEBUG=true
    :param cmd: `class`:`str`
    :param hostname:
    :return:
    """
    if server is None:
        server = 'localhost'
    client = factory.get(server)
    try:
        client.update_image_2(name, image)
    except ContainerNotFound:
        container_ports = []
        if ports:
            ports = parse_list(ports)
            if len(ports) == 1:
                container_ports = [dict(type='tcp', privateport=80, publicport=int(ports[0]))]
            else:
                container_ports = [dict(type='tcp', publicport=int(pub), privateport=int(pri)) for pub, pri in
                                   group_by_2(ports)]

        container_volumes = []
        if volumes:
            container_volumes = [dict(hostvolume=s, containervolume=t) for s, t in group_by_2(parse_list(volumes))]
        if env:
            env = parse_list(env)

        code, result = client.create_container(name, image, hostname=hostname, ports=container_ports,
                                               volumes=container_volumes, env=env,
                                               command=cmd)
        if httplib.OK != code:
            raise Exception("create container failure, code {0}, message: {1}".format(code, result))


def load_settings(src):
    full_path = os.path.join(src, 'matrix.json')
    with open(full_path, 'rb') as f:
        return json.load(f)


def docker_image_name(src='.', release=False, group=None):
    """get building docker image name
    parse matrix.json, and get image:tag

    Example:
        IMAGE_NAME=$(sirius docker_image_name | head -n 1)

    :param src: the dir which container matrix.json, default is current work directory
    :param release: generate release image name
    :param group: project group
    :return:
    """
    settings = load_settings(src)
    name = settings.get('name')

    if group and not name.startswith("{0}/".format(group)):
        name = "{0}/{1}".format(group, name)

    tag = settings.get('release_tag', 'release1') if release else settings.get('tag', 'build1')

    image_name = '{name}:{tag}'.format(name=name, tag=tag)
    print(image_name)
    return image_name


def docker_build_image(workspace=None, matrix_version=None, group=None):
    """build a new image
    Example:
      sirius docker_build_image[:workspace]

    :param workspace: the source code directory, default retrieve workspace from WORKSPACE ENV variable.
    :param matrix_version: the matrix version, default is the new tag
    :param group
    :return:
    """
    if not workspace:
        workspace = os.environ.get('WORKSPACE', '.')

    if not matrix_version:
        matrix_version = os.environ.get('SIRIUS_MATRIX_VERSION')
        if not matrix_version:
            matrix_version = DEFAULT_MATRIX_VERSION
    docker_prepare_build(workspace, group)

    cmd = ('docker run --rm -v {workspace}:/home/matrix '
           '-v /var/run/docker.sock:/var/run/docker.sock docker.neg/matrix:{matrix} /usr/local/bin/matrix.sh')

    local(cmd.format(workspace=workspace, matrix=matrix_version))


def docker_new_build_no(project_slug=None):
    """get new build no

    it's used to build docker image
    Example:
        sirius docker_new_build_no:meerkat | head -n 1

    :param project_slug: project name
    :return:
    """
    url = '{0}/build/{1}'.format(DEIMOS, project_slug)
    response = requests.post(url)
    assert response.status_code == httplib.OK
    obj = response.json()
    print(obj['build_id'])
    return obj['build_id']


def docker_prepare_build(workspace=".", group=None):
    """prepare build docker image

    generate new docker image tag, and rewrite the matrix.json

    :param workspace: the matrix.json location, default './matrix.json'
    :param group
    :return:
    """
    workspace = workspace or '.'

    matrix_json = os.path.join(workspace, 'matrix.json')

    if not os.path.isfile(matrix_json):
        raise ValueError('matrix file is not exists, matrix_json={0}'.format(matrix_json))

    repo = Repo(workspace)
    commit = str(repo.head.commit.hexsha[:5])

    with open(matrix_json, 'rb') as f:
        obj = json.load(f)
        project_slug = obj['name']
        if group and not project_slug.startswith("{0}/".format(group)):
            obj['name'] = "{0}/{1}".format(group, project_slug)

        build_no = docker_new_build_no(project_slug)
        tag = obj['tag']
        version = LooseVersion(tag)
        obj['tag'] = '.'.join(imap(str, chain(version.version[:3], ['build{0}'.format(build_no)])))
        obj['release_tag'] = '.'.join(
            imap(str, chain(version.version[:3], ['release{0}.{1}'.format(build_no, commit)])))

    with open(matrix_json, 'wb') as f:
        json.dump(obj, f)


def docker_release(src='.'):
    image_name = docker_image_name(src)
    release_image_name = docker_image_name(src, release=True)

    # Add new tag
    cmd = 'docker tag -f {0} docker.neg/{1}'.format(image_name, release_image_name)
    local(cmd)

    cmd = 'docker push docker.neg/{0}'.format(release_image_name)
    local(cmd)

    cmd = 'docker rmi docker.neg/{0}'.format(release_image_name)
    local(cmd)


def __deploy(project_name, image, replicas, volumes, env, cmd, hostname, servers, etcd_port=4001):
    for server in servers:
        client = factory.get(server)
        etcd_client = Client(host=server, port=etcd_port)

        for i in xrange(replicas):
            name = "{0}.{1}".format(project_name, i + 1)

            client.delete_container(name)

            container_volumes = []
            if volumes:
                container_volumes = [dict(hostvolume=s, containervolume=t) for s, t in group_by_2(parse_list(volumes))]

            code, result = client.create_container(name, image, hostname=hostname,
                                                   ports=[dict(type='tcp', privateport=8080, publicport=0)],
                                                   volumes=container_volumes, env=env,
                                                   command=cmd)

            if httplib.OK != code:
                raise Exception("create container failure, code {0}, message: {1}".format(code, result))

            code, result = client.get_container(name, True)
            if httplib.OK != code:
                raise Exception("get container information failure, code {0}, message: {1}".format(code, result))

            port = result.NetworkSettings.Ports["8080/tcp"][0].HostPort
            etcd_client.write("/haproxy-discover/services/{0}/upstreams/{1}".format(project_name, name),
                              "{0}:{1}".format(server, port))

        upstreams = etcd_client.get("/haproxy-discover/services/{0}/upstreams".format(project_name))

        for upstream in upstreams.children:
            if isinstance(upstream.key, unicode):
                index = upstream.key[-1:]
                if index and int(index) > replicas:
                    etcd_client.delete(upstream.key)
                    client.delete_container("{0}.{1}".format(project_name, int(index)))
